cloudpathlib/azure/azblobclient.py

```diff
@@ -200,30 +210,30 @@ def _list_dir(
         if prefix and not prefix.endswith("/"):
             prefix += "/"
 
-        yielded_dirs = set()
-
-        # NOTE: Not recursive may be slower than necessary since it just filters
-        #   the recursive implementation
-        for o in container_client.list_blobs(name_starts_with=prefix):
-            # get directory from this path
-            for parent in PurePosixPath(o.name[len(prefix) :]).parents:
-                # if we haven't surfaced this directory already
-                if parent not in yielded_dirs and str(parent) != ".":
-                    # skip if not recursive and this is beyond our depth
-                    if not recursive and "/" in str(parent):
-                        continue
-
-                    yield (
-                        self.CloudPath(f"az://{cloud_path.container}/{prefix}{parent}"),
-                        True,  # is a directory
-                    )
-                    yielded_dirs.add(parent)
-
-            # skip file if not recursive and this is beyond our depth
-            if not recursive and "/" in o.name[len(prefix) :]:
-                continue
-
-            yield (self.CloudPath(f"az://{cloud_path.container}/{o.name}"), False)  # is a file
+        if not recursive:
+            blobs = container_client.walk_blobs(name_starts_with=prefix)
+        else:
+            blobs = container_client.list_blobs(name_starts_with=prefix)
+
+        def is_dir(blob: Union[BlobProperties, BlobPrefix]) -> bool:
+            # walk_blobs returns a BlobPrefix object for directories
+            # https://learn.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.blobprefix?view=azure-python
+            if isinstance(blob, BlobPrefix):
+                return True
+
+            # content_type and content_md5 are never None for files in Azure Blob Storage
+            # if they are None, then the given path is a directory
+            # https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob-properties?tabs=microsoft-entra-id#response-headers
+            return (
+                blob.content_settings.content_md5 is None
+                and blob.content_settings.content_type is None
+            )
+
+        for blob in blobs:
+            # walk_blobs returns folders with a trailing slash
+            blob_path = blob.name.rstrip("/")
```

Comment:
Is it problematic to keep the trailing slashes?

Reply:
I did it to be consistent with the current behaviour, but we can simply not remove the trailing slashes.